File: user-service/src/main/java/com/example/user/client/KakaoApiClient.java
package com.example.user.client;

import com.example.user.dto.KakaoTokenResponse;
import com.example.user.dto.KakaoUserInfoResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

@Component
public class KakaoApiClient {

    private final RestTemplate restTemplate;
    private final String clientId;
    private final String redirectUri;
    private final String tokenUri;
    private final String userInfoUri;

    public KakaoApiClient(
            RestTemplate restTemplate,
            @Value("${kakao.client-id}") String clientId,
            @Value("${kakao.redirect-uri}") String redirectUri,
            @Value("${kakao.token-uri}") String tokenUri,
            @Value("${kakao.user-info-uri}") String userInfoUri) {
        this.restTemplate = restTemplate;
        this.clientId = clientId;
        this.redirectUri = redirectUri;
        this.tokenUri = tokenUri;
        this.userInfoUri = userInfoUri;
    }

    public KakaoTokenResponse getAccessToken(String authCode) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        params.add("grant_type", "authorization_code");
        params.add("client_id", clientId);
        params.add("redirect_uri", redirectUri);
        params.add("code", authCode);

        HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(params, headers);

        return restTemplate.postForObject(tokenUri, request, KakaoTokenResponse.class);
    }

    public KakaoUserInfoResponse getUserInfo(String accessToken) {
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(accessToken);

        HttpEntity<Void> request = new HttpEntity<>(headers);

        return restTemplate.exchange(
                userInfoUri,
                HttpMethod.GET,
                request,
                KakaoUserInfoResponse.class
        ).getBody();
    }
}

File: apps/user-service/src/auth/kakao/kakao-api.client.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import axios from 'axios';
import { KakaoTokenResponse } from './kakao-token-response';
import { KakaoUserInfoResponse } from './kakao-user-info-response';

@Injectable()
export class KakaoApiClient {
  private readonly clientId: string;
  private readonly redirectUri: string;
  private readonly tokenUri: string;
  private readonly userInfoUri: string;

  constructor(private readonly config: ConfigService) {
    this.clientId = config.get<string>('KAKAO_CLIENT_ID', 'test');
    this.redirectUri = config.get<string>('KAKAO_REDIRECT_URI', 'http://localhost');
    this.tokenUri = config.get<string>(
      'KAKAO_TOKEN_URI',
      'https://kauth.kakao.com/oauth/token',
    );
    this.userInfoUri = config.get<string>(
      'KAKAO_USER_INFO_URI',
      'https://kapi.kakao.com/v2/user/me',
    );
  }

  async getAccessToken(authCode: string): Promise<KakaoTokenResponse> {
    const params = new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: this.clientId,
      redirect_uri: this.redirectUri,
      code: authCode,
    });

    const response = await axios.post<KakaoTokenResponse>(
      this.tokenUri,
      params.toString(),
      { headers: { 'Content-Type': 'application/x-www-form-urlencoded' } },
    );

    return response.data;
  }

  async getUserInfo(accessToken: string): Promise<KakaoUserInfoResponse> {
    const response = await axios.get<KakaoUserInfoResponse>(this.userInfoUri, {
      headers: { Authorization: `Bearer ${accessToken}` },
    });

    return response.data;
  }
}
